### ISSUE TITLE
Positionial Hash argument mistaken for keyword argument

### ISSUE THREAD

**Issue by fnordfish** [oracle/truffleruby#2302]
The following method will silently drop the argument `b` on truffleruby 21.0.0,:

```ruby
def foo(a = nil, b = {}, v: false)
  [a, b, v]
end

# truffleruby 21.0.0
foo(:a, {"MISS"=>"ME"})
# => [:a, {}, false]

# ruby 2.7
foo(:a, {"MISS"=>"ME"})
# => [:a, {"MISS"=>"ME"}, false]
```

In my tests, this seems to only be an issue, if there's another default argument before the "Hash" argument.

```ruby
def foo(a, b = {}, v: false)
  [a, b, v]
end

# truffleruby 21.0.0
foo(nil, {"MISS"=>"ME"})
# => [nil, {"MISS"=>"ME"}, false]
```

**Comment by eregon** [comment]
Thank you for the report.
I think what happens is the `{"MISS"=>"ME"}` is considered as keyword arguments (for compatibility with 2.6).
But somehow the extra keyword argument does not cause an error.
In 2.7, keyword argument hashes can contain any key, not just Symbol.

In fact, if the code was `p foo(:a, {:MISS=>"ME"})` then it raises `unknown keyword: :MISS (ArgumentError)` on 2.6, 2.7 and TruffleRuby.

Therefore before Ruby 3 this code is quite brittle, as if there can be Symbol keys in the Hash it will cause such errors.
In Ruby 3 it can be seen as a positional and never keyword argument, and then there is more ambiguity.

Is this code part of some gem, or application or so?
I would be interested to see actual usages of this pattern.

**Comment by fnordfish** [comment]
Thanks for your explanations, that was kind of what I thought was going in.

I was putting together a PR (https://github.com/socketry/console/pull/23/commits/18bf2a4ce7db67b24c6dd268e8f0f7170d3414a2). This gem uses ENV quite a lot for internal default configurations like this `def self.default_log_level(env = ENV)`. I wanted to add a `def self.default_terminal(output = nil, env = ENV, verbose: self.verbose?)` but ended up not being able to overwrite the `env` argument. My workaround/fix was to make `env` also a keyword argument.
Personally, I think having a kw-arg here is cleaner, although in every other method it's just a positional arg - but I can also see people having a different opinion :)


Just to clarify, the sample works in ruby 2.4 and 2.5 the same way.

```ruby
irb(main):001:0> RUBY_VERSION
=> "2.4.9"
irb(main):002:0> def foo(a = nil, b = {}, v: false)
  [a, b, v]
end
=> :foo
irb(main):005:0> foo(:a, {"MISS"=>"ME"})
=> [:a, {"MISS"=>"ME"}, false]
```

```ruby
irb(main):001:0> RUBY_VERSION
=> "2.5.7"
irb(main):002:1* def foo(a = nil, b = {}, v: false)
irb(main):003:1*   [a, b, v]
irb(main):004:0> end
=> :foo
irb(main):005:0> foo(:a, {"MISS"=>"ME"})
=> [:a, {"MISS"=>"ME"}, false]
```

**Comment by eregon** [comment]
I think indeed it's best to avoid having a default positional Hash argument + keyword arguments, that works unreliably across Ruby versions. For instance:
```
~ ∴ ruby -ve 'def foo(a = nil, b = {}, v: false); [a, b, v]; end; p foo(:a, {"MISS"=>"ME", v: true})'
ruby 2.6.6p146 (2020-03-31 revision 67876) [x86_64-linux]
[:a, {"MISS"=>"ME"}, true]
~ ∴ chruby 2.7.2
~ ∴ ruby -ve 'def foo(a = nil, b = {}, v: false); [a, b, v]; end; p foo(:a, {"MISS"=>"ME", v: true})'
ruby 2.7.2p137 (2020-10-01 revision 5445e04352) [x86_64-linux]
-e:1: warning: Splitting the last argument into positional and keyword parameters is deprecated
-e:1: warning: The called method `foo' is defined here
[:a, {"MISS"=>"ME"}, true]
~ ∴ chruby 3.0.0
~ ∴ ruby -ve 'def foo(a = nil, b = {}, v: false); [a, b, v]; end; p foo(:a, {"MISS"=>"ME", v: true})'
ruby 3.0.0p0 (2020-12-25 revision 95aff21468) [x86_64-linux]
[:a, {"MISS"=>"ME", :v=>true}, false]
```

So in 2.7 that's actually a warning if there is a Symbol in Hash, and it causes hash splitting, which is something deprecated in 2.7 and removed in 3.0.

There is still some logic in TruffleRuby to do hash splitting, but it seems to not trigger in this case.
We'll try to take a look.

**Comment by fnordfish** [comment]
totally agree with your conclusions, although my usecase was this:

```console
❯ ruby -ve 'def foo(a = nil, b = {}, v: false); [a, b, v]; end; p foo(:a, {"MISS"=>"ME"}, v: true)'
ruby 2.7.2p137 (2020-10-01 revision 5445e04352) [x86_64-darwin19]
[:a, {"MISS"=>"ME"}, true]
```

which is also valid in ruby 3.0

```console
❯ ruby -ve 'def foo(a = nil, b = {}, v: false); [a, b, v]; end; p foo(:a, {"MISS"=>"ME"}, v: true)'
ruby 3.0.0p0 (2020-12-25 revision 95aff21468) [x86_64-darwin19]
[:a, {"MISS"=>"ME"}, true]
```

**Comment by eregon** [comment]
Yeah, if you pass all 3 arguments it should work fine (that example also works on TruffleRuby).

It's when you want to pass only the 2nd or 3rd it gets quite tricky.
In Ruby <= 2.7, Ruby generally prefers keyword arguments to optional arguments, except in this edge case with optional arguments and non-Symbol keys where it decides to split. If you can have Symbols in the optional argument it wouldn't work (`foo(:a, {:MISS=>"ME"})`).
In Ruby >= 3 it's purely lexical, so it's finally well defined.

**Comment by fnordfish** [comment]
> Yeah, if you pass all 3 arguments it should work fine (that example also works on TruffleRuby).

So true. Totally missed that.

**Comment by fnordfish** [comment]
FYI, it does work in jruby

```
❯ ruby -ve 'def foo(a = nil, b = {}, v: false); [a, b, v]; end; p foo(:a, {"MISS"=>"ME"})'
jruby 9.2.9.0 (2.5.7) 2019-10-30 458ad3e OpenJDK 64-Bit Server VM 15.0.2+7 on 15.0.2+7 +jit [darwin-x86_64]
[:a, {"MISS"=>"ME"}, false]
```

**Comment by eregon** [comment]
Fixed in d8890e7f55004c6ae6c55e0baf013501e4820adc.
